### ISSUE TITLE
Rendering a card is "brutal" - is there a more elegant way?

### ISSUE THREAD

**Issue by ntoll** [pyscript/pypercard#35]
In this proof-of-concept, to render a card we either take a string or the `innerHTML` of a `<template>` tag in the DOM, then destroy the current content of the `<pyper-app>` tag and emit the resulting element that represents the new card into the tag. I wonder if we could just change the `display: none` on pre-rendered cards? It would make transition effects easier to do too.

**Comment by ntoll** [comment]
Ideas welcome.

**Comment by WebReflection** [comment]
let me start saying that a `<templte>` element doesn't need to provide `innerHTML` rather its `content` which is already a pre-available document fragment with the content the `template` provides so that there's no need to use innerHTML ... but here it's how I am going to demo today my stuff:

```js
const templates = [...document.querySelectorAll('body > template')];

const showCard = (id = templates[0].id) => {
  document.body.replaceChildren(
    templates.find(({id: tid}) => tid === id).content.cloneNode(true)
  );
  for (const button of document.body.querySelectorAll('[transition]')) {
    button.addEventListener('click', nextCard);
  }

  function nextCard({currentTarget}) {
    showCard(currentTarget.getAttribute('transition'));
  }
};

showCard();
```

This is a simplified code for a simplified demo but the logic to swap templates is there ... maybe it helps?

**Comment by WebReflection** [comment]
P.S. of course you can avoid the `replaceChildren` dance and use a `display: none` or a `hidden = true` on the container so if you want to bootstrap out of the box all templates and make only the starting one visible and then orchestrate the rest is also possible. Let me know if you need any help there.

**Comment by WebReflection** [comment]
P.S.2 ... if you don't want to ever see errors in console due duplicated ids (remember this topic) you still need to replace the content of the page otherwise IDs will backfire on the logic so that replacing each card each time is kinda mandatory to avoid conflicts with IDs used in multiple cards (if these are the same).

Chrome warns you there are issues there in console, I expect other browsers too. Actually the editor might trigger those warnings if two different cards have the same button ID, as example. It's not an issue for the editor, but it's hard to explain or reason about for our users, imho.

**Comment by ntoll** [comment]
This helps a lot. The important thing is the rather shonky code we have right now points the way to how it could be done properly. The missing aspect for me is to identify places in the template into which values are inserted by the code. Currently this is just a Python string `format` function on the `innerHTML`. So `"Hello {name}".format(name="Fred")` becomes `"Hello Fred" when rendered. I realise this is not a good / performant nor idiomatic way to do things (but it's "good enough" :tm: given impending PyCon).

I'd rather like to avoid `replaceChildren` dance and render all the cards with the correct `display` value. I also realise it makes `id` a problem. But perhaps instead of `id` for elements within a card, we use `name` instead and get Python to enforce the uniqueness of the `name` within the context of a specific card..?

Anyway, I feel this can be improved a lot, but it's definitely post-PyCon for now as we just "make it work" :tm:. 

"Make it work properly" is a feeling that I feel keenly, but I'm sitting on my hands until after PyCon. ;-)

**Comment by WebReflection** [comment]
the format dance can be orchestrated in a better way for sure, my hints were around how template element works ... it's not "*yet another container*" it has tons of features in it out of the box.

IDs to me can be simply automagically created via the card name / identifier so that a button with name/id `do_stuff` becomes `card_1_do_stuff` and we're good to go. This should be (imho) transparent to users while editing IDs, but we might warn them if a single card contains duplicated IDs for two different buttons, inputs, text, whatsoever.

**Comment by ntoll** [comment]
@WebReflection totally agree WRT the ids are transparent to the user. This speaks to @mchilvers's point about the user's layer of things being a step away from the DOM (currently this is not the case). Having this separation would help us keep things simple for beginner users.

**Comment by ntoll** [comment]
Addressed (fixed?) in #37
